### DISCUSSION TITLE
how to install multiple instances of a formula?

### DISCUSSION THREAD

**Discussion by furrnace** [Homebrew/discussions#4070] [Everyday usage]
greetings, I would like to install 2 differently parameterized instances of a given formula, so that each instance has a different name and each may be used independently of each other, i.e. there is no need to uninstall the formula and to install it newly with different parameters.

can this be achieved? f so, a pointer to a document describing this would be appreciated.

**Comment by SMillerDev** [comment]
No, formula are unique by name. If you want to install the same software twice you will need to make a copy with a new name.

**Reply by furrnace** [reply]
what to you mean by "copy with a new name"? how would I do this in homebrew (do I find this in the documentation, under which chapter would that be?)

**Reply by SMillerDev** [reply]
https://docs.brew.sh/Manpage#extract---version---force-formula-tap would probably allow you to make a copy. Otherwise you can just manually copy the formula file.

**Comment by furrnace** [comment]
to make sure I understood that, is this what I need to do to have 2 instances of formula example named example-a and example-b:

brew tap my/example-a https://github.com/repouser/reponame
brew tap my/example-b https://github.com/repouser/reponame

and then to have both instances installed simultaneously, I would need to rename the formula, too, not just the tap?
how do I do this? Is it enough to just rename the .rb file or do I also need to rename the class in the first line of the .rb file? Anything else I need to do for this to work properly?

cd /usr/local/Homebrew/Library/Taps/my/homebrew-example-a
mv example.rb example-a.rb

cd /usr/local/Homebrew/Library/Taps/my/homebrew-example-b
mv example.rb example-b.rb

Questions that come to my mind (as I have zero experience with maintaining formulae):
-if the formula installs several programs, how can I make sure the program file names for instances example-a and example-b are different?
-do  changes I make to the .rb file (by renaming and/or editing its content) endure a "brew upgrade --fetch HEAD my/example-a/example-a" ?

**Reply by SMillerDev** [reply]
You don't need different taps. Taps are just lists of formulae. You do need two different formulae though. That means different files and classes.

**Reply by furrnace** [reply]
I created new taps because the exctract command would not create a duplicate formula in the same tap.

How will the different formulae names lead to different names for the binaries? do I have to do something so that the binary names or the names of the links pointing to the binaries do not clash?

**Reply by SMillerDev** [reply]
That depends on the software. I can't tell you in this hypothetical scenario.

**Comment by furrnace** [comment]
lets assume we want two different instances of RfidResearchGroup/homebrew-proxmark3, one using "--with-generic" and one using "--with-blueshark"

After creating 2 new taps using the extract command I used the edit command to change the class name in both proxmark3.rb files to "Proxmark3-GENERIC" and "Proxmark3-RDV4", respectively. Also I renamed the proxmark3.rb files to proxmark3-generic.rb and proxmark3-rdv4.rb, respectively.

After that, the install command wouldn't work and complain that the options "--with-generic" or "--with-blueshark" are unknown. Obviously I missed to change something else, but I don't know what that would be.

**Reply by gromgit** [reply]
> After creating 2 new taps using the extract command I used the edit command to change the class name in both proxmark3.rb files to "Proxmark3-GENERIC" and "Proxmark3-RDV4", respectively. Also I renamed the proxmark3.rb files to proxmark3-generic.rb and proxmark3-rdv4.rb, respectively.

At the very least, you've named the classes wrong. From https://docs.brew.sh/Formula-Cookbook#a-quick-word-on-naming:
> Filenames should be all lowercase, and class names should be the strict CamelCase equivalent, e.g. formulae `gnu-go` and `sdl_mixer` become classes `GnuGo` and `SdlMixer`, even if part of their name is an acronym.

That said, the [original formula](https://github.com/RfidResearchGroup/homebrew-proxmark3/blob/master/proxmark3.rb) already defines both options, so you're probably doing something else wrong too. What are:
* the _exact_ contents of your `proxmark3-generic.rb` and `proxmark3-rdv4.rb`,
* the _exact_ install command you typed, and
* the _exact_ output you saw?

**Reply by furrnace** [reply]
(see in one of my earlier posts:) I renamed the original file proxmark3.rb and the class within the file. From this follows that the content is identical with the original proxmark3.rb (except for the class name).

these are the commands used for the rdv4 instance, the commands for the generic instance are the same, just replace rdv4 with generic:
```
brew tap my/pm3rdv4 https://github.com/RfidResearchGroup/homebrew-proxmark3
brew edit --formula my/pm3rdv4/proxmark3
```
in line 1 change class name from `Proxmark3` to `Proxmark3-RDV4`  (and this should probably be Proxmark3Rdv4)
```
mv /usr/local/Homebrew/Library/Taps/my/homebrew-pm3rdv4/proxmark3.rb /usr/local/Homebrew/Library/Taps/my/homebrew-pm3rdv4/proxmark3-rdv4.rb
brew install --HEAD --with-blueshark my/pm3rdv4/proxmark3-rdv4
```

The output:
```
Running `brew update --auto-update`...
==> Auto-updated Homebrew!
==> Updated Homebrew from 3.6.16 (e76c55e18) to 3.6.17 (015ee67a7).
Updated 5 taps (homebrew/cask-versions, homebrew/command-not-found, homebrew/core, homebrew/cask and homebrew/services).

You have 9 outdated formulae and 1 outdated cask installed.
You can upgrade them with brew upgrade
or list them with brew outdated.


The 3.6.17 changelog can be found at:
  https://github.com/Homebrew/brew/releases/tag/3.6.17
Usage: brew install [options] formula|cask [...]

Install a formula or cask. Additional options specific to a formula may be
appended to the command.

Unless HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK is set, brew upgrade or brew
reinstall will be run for outdated dependents and dependents with broken
linkage, respectively.

Unless HOMEBREW_NO_INSTALL_CLEANUP is set, brew cleanup will then be run for
the installed formulae or, every 30 days, for all formulae.

Unless HOMEBREW_NO_INSTALL_UPGRADE is set, brew install formula will
upgrade formula if it is already installed but outdated.

  -d, --debug                      If brewing fails, open an interactive
                                   debugging session with access to IRB or a
                                   shell inside the temporary build directory.
  -f, --force                      Install formulae without checking for
                                   previously installed keg-only or non-migrated
                                   versions. When installing casks, overwrite
                                   existing files (binaries and symlinks are
                                   excluded, unless originally from the same
                                   cask).
  -v, --verbose                    Print the verification and postinstall steps.
  -n, --dry-run                    Show what would be installed, but do not
                                   actually install anything.
      --formula, --formulae        Treat all named arguments as formulae.
      --ignore-dependencies        An unsupported Homebrew development flag to
                                   skip installing any dependencies of any kind.
                                   If the dependencies are not already present,
                                   the formula will have issues. If you're not
                                   developing Homebrew, consider adjusting your
                                   PATH rather than using this flag.
      --only-dependencies          Install the dependencies with specified
                                   options but do not install the formula
                                   itself.
      --cc                         Attempt to compile using the specified
                                   compiler, which should be the name of the
                                   compiler's executable, e.g. gcc-7 for GCC
                                   7. In order to use LLVM's clang, specify
                                   llvm_clang. To use the Apple-provided
                                   clang, specify clang. This option will only
                                   accept compilers that are provided by
                                   Homebrew or bundled with macOS. Please do not
                                   file issues if you encounter errors while
                                   using this option.
  -s, --build-from-source          Compile formula from source even if a
                                   bottle is provided. Dependencies will still
                                   be installed from bottles if they are
                                   available.
      --force-bottle               Install from a bottle if it exists for the
                                   current or newest version of macOS, even if
                                   it would not normally be used for
                                   installation.
      --include-test               Install testing dependencies required to run
                                   brew test formula.
      --HEAD                       If formula defines it, install the HEAD
                                   version, aka. main, trunk, unstable, master.
      --fetch-HEAD                 Fetch the upstream repository to detect if
                                   the HEAD installation of the formula is
                                   outdated. Otherwise, the repository's HEAD
                                   will only be checked for updates when a new
                                   stable or development version has been
                                   released.
      --keep-tmp                   Retain the temporary files created during
                                   installation.
      --debug-symbols              Generate debug symbols on build. Source will
                                   be retained in a cache directory. 
      --build-bottle               Prepare the formula for eventual bottling
                                   during installation, skipping any
                                   post-install steps.
      --bottle-arch                Optimise bottles for the specified
                                   architecture rather than the oldest
                                   architecture supported by the version of
                                   macOS the bottles are built on.
      --display-times              Print install times for each package at the
                                   end of the run.
  -i, --interactive                Download and patch formula, then open a
                                   shell. This allows the user to run
                                   ./configure --help and otherwise determine
                                   how to turn the software package into a
                                   Homebrew package.
  -g, --git                        Create a Git repository, useful for creating
                                   patches to the software.
      --overwrite                  Delete files that already exist in the prefix
                                   while linking.
      --cask, --casks              Treat all named arguments as casks.
      --[no-]binaries              Disable/enable linking of helper executables
                                   (default: enabled).
      --require-sha                Require all casks to have a checksum.
      --[no-]quarantine            Disable/enable quarantining of downloads
                                   (default: enabled).
      --adopt                      Adopt existing artifacts in the destination
                                   that are identical to those being installed.
                                   Cannot be combined with --force.
      --skip-cask-deps             Skip installing cask dependencies.
      --zap                        For use with brew reinstall --cask. Remove
                                   all files associated with a cask. May remove
                                   files which are shared between applications.
      --appdir                     Target location for Applications (default:
                                   /Applications).
      --colorpickerdir             Target location for Color Pickers (default:
                                   ~/Library/ColorPickers).
      --prefpanedir                Target location for Preference Panes
                                   (default: ~/Library/PreferencePanes).
      --qlplugindir                Target location for QuickLook Plugins
                                   (default: ~/Library/QuickLook).
      --mdimporterdir              Target location for Spotlight Plugins
                                   (default: ~/Library/Spotlight).
      --dictionarydir              Target location for Dictionaries (default:
                                   ~/Library/Dictionaries).
      --fontdir                    Target location for Fonts (default:
                                   ~/Library/Fonts).
      --servicedir                 Target location for Services (default:
                                   ~/Library/Services).
      --input-methoddir            Target location for Input Methods (default:
                                   ~/Library/Input Methods).
      --internet-plugindir         Target location for Internet Plugins
                                   (default: ~/Library/Internet Plug-Ins).
      --audio-unit-plugindir       Target location for Audio Unit Plugins
                                   (default:
                                   ~/Library/Audio/Plug-Ins/Components).
      --vst-plugindir              Target location for VST Plugins (default:
                                   ~/Library/Audio/Plug-Ins/VST).
      --vst3-plugindir             Target location for VST3 Plugins (default:
                                   ~/Library/Audio/Plug-Ins/VST3).
      --screen-saverdir            Target location for Screen Savers (default:
                                   ~/Library/Screen Savers).
      --language                   Comma-separated list of language codes to
                                   prefer for cask installation. The first
                                   matching language is used, otherwise it
                                   reverts to the cask's default language. The
                                   default value is the language of your system.
  -q, --quiet                      Make some output more quiet.
  -h, --help                       Show this message.
Error: invalid option: --with-blueshark
```

**Reply by SMillerDev** [reply]
But if you want to edit an install you need to edit the formula file. Arguments you pass to brew should only be arguments brew understands.

**Reply by furrnace** [reply]
Well, I thought I did that by editing the .rb file (that was also what I was asking way higher up, referring to the .rb files). 
If I shouldn't edit that, which file(s) should I edit and rename instead?

**Reply by SMillerDev** [reply]
Well, no matter what you edit. It won't add additional commands line options to brew.

**Reply by gromgit** [reply]
> in line 1 change class name from `Proxmark3` to `Proxmark3-RDV4` (and this should probably be Proxmark3Rdv4)

There's no "probably" about it. Failure to adhere to Homebrew's class-naming conventions will result in Homebrew not being able to find your formula's class, and therefore not be able to do anything with it.

Fix that, and see if you can make any progress.

**Reply by furrnace** [reply]
@gromgit Thank you your comment was the actual helpful one!

I now have both, proxmark3-rdv4 and proxmark3-generic, compiled and proxmark3-rdv4 is linked. I still would have to brew link --overwrite proxymark3-rdv4 or proxmark3-generic to switch between versions. As this is not really faster than doing a brew reinstall with different parameters - I haven't gained much yet.

The next step I think would be to use some brew mechanism(s) to link both formulae, proxmark3-rdv4 and proxmark3-generic, to different names in /usr/local/bin so that both are readily available in the execution path.

What is there when proxmark3-rdv4 is linked:

/usr/local/bin/pm3 -> /usr/local/Cellar/proxmark3-rdv4/HEAD-53808f2/bin/pm3
/usr/local/bin/pm3-flash -> /usr/local/Cellar/proxmark3-rdv4/HEAD-53808f2/bin/pm3-flash
/usr/local/bin/pm3-flash-all -> /usr/local/Cellar/proxmark3-rdv4/HEAD-53808f2/bin/pm3-flash-all
/usr/local/bin/pm3-flash-bootrom -> /usr/local/Cellar/proxmark3-rdv4/HEAD-53808f2/bin/pm3-flash-bootrom
/usr/local/bin/pm3-flash-fullimage -> /usr/local/Cellar/proxmark3-rdv4/HEAD-53808f2/bin/pm3-flash-fullimage
/usr/local/bin/proxmark3 -> /usr/local/Cellar/proxmark3-rdv4/HEAD-53808f2/bin/proxmark3

how could I make brew link proxmar3-rdv4 to something like /usr/local/bin/pm3-rdv4 instead of /usr/local/bin/pm3 (and likewise for proxmark3-generic: /usr/local/bin/pm3-generic instead of /usr/local/bin/pm3) ?

**Reply by gromgit** [reply]
You'd have to rename the binaries during the build itself, something like this (UNTESTED):
```ruby
Dir.glob("#{bin}/*") do |f|
  File.rename "#{f}" "#{f}-rdv4"
end
```

**Reply by furrnace** [reply]
OK, currently I don't know yet, where in the .rb file the build section is. I am willing to learn enough about this to be able to modify the files and went to the documentation on brew.sh. To be honest, I didn't find a section that provided an entry point that was suitable for me to get an understanding of the structure of the .rb files and which "sections" are there and how to recognize them. Certainly that's my fault. I would be happy if someone could point me to a description/documentation that gives an overview of the structure of .rb files.
